cli/cli/helpers/grafloki/kubernetes_grafloki.go

```diff
@@ -0,0 +1,477 @@
+	glNamespace, err := k8sManager.CreateNamespace(ctx, graflokiNamespace, map[string]string{}, map[string]string{})
+	shouldRemoveNamespace := false
+	defer func() {
+		if shouldRemoveNamespace {
+			if err := k8sManager.RemoveNamespace(ctx, glNamespace); err != nil {
+				logrus.Warnf("Attempted to remove namespace '%v' after an error occurred creating it but an error occurred removing it.", graflokiNamespace)
+				logrus.Warnf("Manually remove namepsace with name: %v", glNamespace.Name)
+			}
+		}
+	}()
+	if err != nil {
+		return "", stacktrace.Propagate(err, "An error occurred creating namespace '%v'", graflokiNamespace)
+	}
+
+	lokiDeployment, err := k8sManager.CreateDeployment(
+		ctx,
+		graflokiNamespace,
+		lokiDeploymentName,
+		lokiLabels,
+		map[string]string{},
+		[]apiv1.Container{}, // no init containers
+		[]apiv1.Container{
+			{
+				Name:  "loki",
+				Image: lokiImage,
+				Ports: []apiv1.ContainerPort{
+					{
+						Name:          "",
+						HostPort:      0,
+						ContainerPort: lokiPort,
+						Protocol:      "",
+						HostIP:        "",
+					},
+				},
+				Command:    nil,
+				Args:       nil,
+				WorkingDir: "",
+				EnvFrom:    nil,
+				Env:        nil,
+				Resources: apiv1.ResourceRequirements{
+					Limits:   nil,
+					Requests: nil,
+					Claims:   nil,
+				},
+				ResizePolicy:  nil,
+				VolumeMounts:  nil,
+				VolumeDevices: nil,
+				LivenessProbe: nil,
+				ReadinessProbe: &apiv1.Probe{
+					ProbeHandler: apiv1.ProbeHandler{
+						Exec: nil,
+						HTTPGet: &apiv1.HTTPGetAction{
+							Path:        "/ready",
+							Port:        intstr.FromInt(lokiPort),
+							Host:        "",
+							Scheme:      "",
+							HTTPHeaders: nil,
+						},
+						TCPSocket: nil,
+						GRPC:      nil,
+					},
+					InitialDelaySeconds:           lokiProbInitialDelaySeconds,
+					TimeoutSeconds:                lokiProbeTimeoutSeconds,
+					PeriodSeconds:                 lokiProbePeriodSeconds,
+					SuccessThreshold:              0,
+					FailureThreshold:              0,
+					TerminationGracePeriodSeconds: nil,
+				},
+				StartupProbe:             nil,
+				Lifecycle:                nil,
+				TerminationMessagePath:   "",
+				TerminationMessagePolicy: "",
+				ImagePullPolicy:          "",
+				SecurityContext:          nil,
+				Stdin:                    false,
+				StdinOnce:                false,
+				TTY:                      false,
+			},
+		},
+		[]apiv1.Volume{},
+		&apiv1.Affinity{
+			NodeAffinity:    nil,
+			PodAffinity:     nil,
+			PodAntiAffinity: nil,
+		})
+	if err != nil {
+		return "", stacktrace.Propagate(err, "An error occurred creating Loki deployment.")
+	}
+	shouldRemoveLokiDeployment := false
+	defer func() {
+		if shouldRemoveLokiDeployment {
+			if err := k8sManager.RemoveDeployment(ctx, graflokiNamespace, lokiDeployment); err != nil {
+				logrus.Warnf("Attempted to remove Loki deployment after an error occurred creating it but an error occurred removing it.")
```

Comment:
Technically there's not an error creating the deployment, right? It's actually that an unrelated error happened afterwards, and to clean up things we tried to remove the deployment (but then an error happened with _that_)